orc8r/gateway/python/magma/ctraced/trace_manager.py

```diff
@@ -105,42 +119,120 @@ def end_trace(self) -> EndTraceResult:
-                    break
-                logging.info("TraceManager: Waiting 1s for trace file to be "
-                             "written...")
-                time.sleep(1)
+            self._wait_until_trace_file_exists()
 
         # Perform postprocessing of capture file with TShark display filters
-        if len(self._display_filters) > 0:
-            succeeded = self._postprocess_trace()
-            if not succeeded:
-                return EndTraceResult(False, None)
-
-            # Read trace data into bytes
-            with open(self._trace_filename_postprocessed, "rb") as trace_file:
-                data = trace_file.read()  # type: bytes
-        else:
-            # Read trace data into bytes
-            with open(self._trace_filename, "rb") as trace_file:
-                data = trace_file.read()  # type: bytes
-
-        # Ensure the tmp trace file is deleted
-        self._ensure_tmp_file_deleted()
-        self._trace_filename = ""
+        succeeded = self._conditionally_postprocess_trace()
+        if not succeeded:
+            return EndTraceResult(False, None)
 
-        self._is_active = False
+        data = self._get_final_trace_data() # type: bytes
 
+        self._cleanup_trace()
         logging.info("TraceManager: Call trace has ended")
 
         # Everything cleaned up, return bytes
         return EndTraceResult(True, data)
 
+    def _execute_start_trace_command(self, command: List[str]) -> bool:
+        """Executes a command to start a call trace
+
+        Args:
+            command: Shell command with each token ordered in the list.
+              example: ["tshark", "-i", "eth0"] would be for "tshark -i eth0"
+
+        Returns:
+            True if successfully executed command
+        """
+        logging.info("TraceManager: Starting trace with %s, command: [%s]",
+                     self._tool_name, ' '.join(command))
+
+        self._ensure_trace_directory_exists()
+
+        # TODO(andreilee): Handle edge case where only one instance of the
+        #                  process can be running, and may have been started
+        #                  by something external as well.
+
+
+        # TODO(andreilee): Make sure that a fast failure is detected
+        def run_tracing_in_thread(on_exit, command: List[str]):
+            try:
+                self._proc = subprocess.Popen(
+                    command,
+                    stdout=subprocess.PIPE,
+                    stderr=subprocess.STDOUT)
+            except SubprocessError as e:
+                logging.error("TraceManager: Failed to start trace: %s",
+                              str(e))
+
+            self._is_active = True
+            logging.info("TraceManager: Successfully started trace with %s",
+                         self._tool_name)
+
+            self._proc.wait()
+            on_exit()
+            return
+
+        thread = threading.Thread(target=run_tracing_in_thread,
```

Comment:
Why are we running the subprocess in a thread ?

Reply:
Just following the advice [here](https://stackoverflow.com/questions/2581817/python-subprocess-callback-when-cmd-exits) from stackoverflow.